faunadb/stream_subscription.go

```diff
@@ -0,0 +1,83 @@
+package faunadb
+
+import "sync"
+
+func newSubscription(client *FaunaClient, query Expr, config ...StreamConfig) StreamSubscription {
+	sub := StreamSubscription{
+		query,
+		streamConfig{
+			[]StreamField{},
+		},
+		client,
+		streamConnectionStatus{status: StreamConnIdle},
+		make(chan StreamEvent),
+	}
+	for _, fn := range config {
+		fn(&sub)
+	}
+	return sub
+}
+
+type streamConnectionStatus struct {
+	mu     sync.Mutex
+	status StreamConnectionStatus
+}
+
+func (s *streamConnectionStatus) Set(status StreamConnectionStatus) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.status = status
+}
+
+func (s *streamConnectionStatus) Get() StreamConnectionStatus {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.status
+}
+
+// StreamSubscription dispatches events received to the registered listener functions.
+// New subscriptions must be constructed via the FaunaClient stream method.
+type StreamSubscription struct {
+	query    Expr
+	config   streamConfig
+	client   *FaunaClient
+	status   streamConnectionStatus
+	messages chan StreamEvent
+}
+
+// Query returns the query used to initiate the stream
+func (sub *StreamSubscription) Query() Expr {
+	return sub.query
+}
+
+// Status returns the current stream status
+func (sub *StreamSubscription) Status() StreamConnectionStatus {
+	return sub.status.Get()
+}
+
+// Start initiates the stream subscription.
+func (sub *StreamSubscription) Start() error {
+	return sub.client.startStream(sub)
+}
+
+func isClosed(ch <-chan StreamEvent) bool {
+	select {
+	case <-ch:
+		return true
+	default:
+	}
+
+	return false
+}
+
+// Close eventually closes the stream
+func (sub *StreamSubscription) Close() {
+	if !isClosed(sub.messages) {
```

Comment:
What's the relationship between the `messages` and the closed status of a subscription? Can you rely on the `status` variable instead?

Reply:
We need to close event channel in case we are going to close subscription.

Reply:
My point is: I don't think this `isClosed` function respects Golang channel semantics: https://play.golang.org/p/gXiR4_5fW8D

As I was referring to in the client receive loop, I think we need a close signal (channel) for this propose.

Reply:
@erickpintor 
my idea with sending close signal was the following:
```
1:for {
2:	select {
3:	case <- subscription.closeSignal:
4:		subscription.status.Set(StreamConnClosed)
5:		close(subscription.messages)
6:	default:
7:		if val, err := client.parseResponse()...
8:}
```

Here is the issue 
when we obtain data we:
	- parsing (line 7)
	- event generated (this we are going to send closing signal)
	- we are going to parse new portion of data (line 7)
	- process logic inside Error event: in this case we are sending close signal 
but this signal is never consumed by line 3 since we are at line 7 waiting for new chunk of data

Any idea how to handle this ? or we can just rewrite isClosed since it works.

Reply:
Pls ignore my previous comment and check new implementation.